File: esami-vecchi/Gen19/Esame21-01-19.py
'''integrale della sfera n dimensionale con metodo monte carlo'''

from ROOT import *
import math as m

rnd=TRandom2();

l=10					##semi-lato cubo n dim
r=l				##raggio crf n dim
dim=4			###Volume=alpha*r**dim
norm=(2*l)**dim/(r**dim)
N=5000000
fav=0.0
alpha=0.0
atteso=m.pi**(dim/2)/TMath.Gamma(dim/2+1)
delta=0.0
print "valore atteso per alpha", atteso

##TF1("f","TMath::Constant()",0,N)
gr=TGraph()
gr.SetTitle("grafico")
gr.GetXaxis().SetTitle("N")
gr.GetYaxis().SetTitle("alpha")

rnd.SetSeed(123456789)

for i in range(1,N):
	x=l-2*l*rnd.Rndm()			##le coordinate del punto devono appartenere allo spazio campionario 
	y=l-2*l*rnd.Rndm()					##il quadrato n dim [-l,l]
	z=l-2*l*rnd.Rndm()
	w=l-2*l*rnd.Rndm()
	d=x**2+y**2+z**2+w**2
	#print x,y,z,w

	if d<=r**2:
		fav=fav+1
		alpha=(fav/N)*norm
		delta=atteso-alpha
		if(i%1000000==0):
			print alpha
			gr.SetPoint(gr.GetN(),i,alpha)

print alpha

c=TCanvas()
gr.Draw("AP")
gApplication.Run(True)
		
		

File: esami-vecchi/Gen19/esame21-01-19.cpp
#include <TGraph.h>
#include <TF1.h>
#include <TCanvas.h>
#include <TH1D.h>
#include <TApplication.h>
#include <TRandom2.h>
#include <TF1.h>
#include <iostream>
#include <cmath>

using namespace std;

int main(){

TApplication app("app",0,NULL);

TRandom2 rnd;

double l = 1;					
double r=l	;			
int dim=4	;				
double norm=pow((2*l),dim)/pow(r,dim);
double N=1E09;
double fav=0.0;
double alpha=0.0;
double atteso=pow(TMath::Pi(),dim/2)/TMath::Gamma(dim/2+1);
double delta=0.0;
cout << "valore atteso per alpha " << atteso << endl << endl;
double x,y,z,w;
 
TGraph gr;
gr.SetTitle("grafico");
gr.GetXaxis()->SetTitle("N");
gr.GetYaxis()->SetTitle("alpha");

rnd.SetSeed(12345);

for(int i=1;i<N;i++){
	x=l-2*l*rnd.Rndm()	;		
	y=l-2*l*rnd.Rndm()	;				
	z=l-2*l*rnd.Rndm()	;
	w=l-2*l*rnd.Rndm()	;
	double d=x*x+y*y+z*z+w*w	;
	

	if (d<=r*r){
		fav=fav+1;
		alpha=(fav/N)*norm;
		delta=atteso-alpha;
		if(i%1000000==0){
			cout <<  alpha << endl;
			gr.SetPoint(gr.GetN(),i,alpha);
		}
	}
}

cout <<  alpha << endl;
cout << gr.GetN() << endl;
TCanvas *c1= new TCanvas("c1","Grafico",10,10,500,500);
c1->SetLogx();
gr.Draw("AP");
app.Run(true);

	return 0;
}

